plugins/trace_api_plugin/include/eosio/trace_api_plugin/store_provider.hpp

```diff
@@ -0,0 +1,186 @@
+
+      const boost::filesystem::path _slice_dir;
+      const uint32_t _width;
+
+      static constexpr uint32_t _current_version = 1;
+
+      static constexpr uint _slice_size = 10;
+      static constexpr uint _null_term_size = 1;
+      static constexpr const char* _trace_prefix = "trace_";
+      static constexpr const char* _trace_index_prefix = "trace_index_";
+      static constexpr uint _max_file_size = 12 + 10 + 1 + 10 + 4 + 1; // "trace_index_" + 10-digits + '-' + 10-digits + ".log" + null-char
+   };
+
+   class store_provider {
+   public:
+      store_provider(const boost::filesystem::path& slice_dir, uint32_t width);
+
+      void append(const block_trace_v0& bt);
+      void append_lib(uint32_t lib);
+
+      /**
+       * Read the metadata log font-to-back starting at an offset passing each entry to a provided functor/lambda
+       *
+       * @tparam Fn : type of the functor/lambda
+       * @param block_height : height of the requested data
+       * @param offset : initial offset to read from
+       * @param fn : the functor/lambda
+       * @return the highest offset read during this scan
+       */
+      template<typename Fn>
+      uint64_t scan_metadata_log_from( uint32_t block_height, uint64_t offset, Fn&& fn ) {
+         // ignoring offset
+         offset = 0;
+         fc::cfile index;
+         const uint32_t slice_number = _slice_provider.slice_number(block_height);
+         _slice_provider.find_index_slice(slice_number, false, index);
+         const uint64_t end = file_size(index.get_file_path());
+         offset = index.tellp();
+         uint64_t last_read_offset = offset;
+         bool cont = true;
+         while (offset < end && cont) {
+            const auto metadata = extract_store<metadata_log_entry>(index);
+            cont = fn(metadata);
+            last_read_offset = offset;
+            offset = index.tellp();
+         }
+         return last_read_offset;
+      }
+
+      /**
+       * Read from the data log
+       * @param block_height : the block_height of the data being read
+       * @param offset : the offset in the datalog to read
+       * @return empty optional if the data log does not exist, data otherwise
+       * @throws std::exception : when the data is not the correct type or if the log is corrupt in some way
+       *
+       */
+      std::optional<data_log_entry> read_data_log( uint32_t block_height, uint64_t offset ) {
+         const uint32_t slice_number = _slice_provider.slice_number(block_height);
+         fc::cfile trace;
+         if (_slice_provider.find_trace_slice(slice_number, false, trace)) {
+            const std::string offset_str = boost::lexical_cast<std::string>(offset);
+            const std::string bh_str = boost::lexical_cast<std::string>(block_height);
+            throw malformed_slice_file("Requested offset: " + offset_str + " to retrieve block number: " + bh_str + " but this trace file is new, so there are no traces present.");
```

Comment:
Pending a discussion, we need to decide whether this should return an empty `optional` or `throw`. 

If it remains as a `throw` the message should be that the "trace file does not exist" right?  This reader should not create a new one.